RFC-0001-Agile Coretime.md

```diff
@@ -0,0 +1,291 @@
+
+### Broker-chain
+
+The Broker-chain is a new system parachain. It has the responsibility of providing the Relay-chain via UMP with scheduling information of:
+
+- How many cores which should be made available during the next session.
+- Which cores should be allocated to which para IDs.
+
+Any cores left unallocated are assumed by the Broker-chain to be used for Instantaneous Coretime sales.
+
+It is also expected to receive information from the Relay-chain on total revenue from instantaneous coretime sales on a per-block basis.
+
+### Bulk Sales
+
+There is a periodic sale of Coretime happening at a period of `BULK_PERIOD`.  A number of instances of Coretime spanning from `LEADIN_PERIOD` in the future to `LEADIN_PERIOD + BULK_PERIOD` in the future of a single Polkadot UC Core is offered by the Polkadot System at a fixed price.
+
+These instances which are owned by purchaser are called *regions*. This sale happens on the Broker-chain. Regions are quantized into atomic periods of `TIMESLICE`, into which `BULK_PERIOD` divides a whole number of times. The `Timeslice` type is a `u16` which can be multiplied by `TIMESLICE` to give a `BlockNumber` value indicating the same period in times of (Relay-chain) blocks.
+
+The Broker-chain aims to sell some `BULK_TARGET` of Cores, up to some `BULK_LIMIT`. It makes this sale in a single batch `LEADIN_PERIOD` prior to the beginning of the period being sold. The Broker chain publishes a price for this batch of sales for the `BULK_TARGET` period prior to the sale execution.
+
+Accounts may call a `purchase` function with the appropriate funds in place to have their funds reserved and signal an intention to purchase Bulk Coretime for the forthcoming period. One account may have only one pending purchase. Any number of accounts may attempt to `purchase` Bulk Coretime for the forthcoming period, but the order is recorded.
+
+If there are more purchase requests than available cores for purchase in this period, then requests are processed in incoming order. Any additional purchase orders are carried over but marked as such. A purchase is only cancellable if it was carried over.
+
+When a region of Bulk Coretime is initially issued through this purchase, the price it was purchased for is recorded, in addition to the beginning and end Relay-chain block numbers to which it corresponds.
+
+The Broker-chain SHALL record this information in a storage double-map called Regions, keyed first by the current bulk `SaleIndex` (a `u16` starting at zero and incrementing with each sale), then secondarily by a `RegionId`. It shall map into a value of `RegionRecord`:
+
+```rust
+type SaleIndex = u16;
+struct RegionId {
+    core: CoreIndex,  // A `u16`.
+    begin: Timeslice, // A `u16`.
+}
+struct RegionRecord {
+    owner: AccountId,
+    end: Timeslice,
+    price: Option<Balance>,
+    allocation: Option<Vec<ParaId>>, // begins set to `None`
```

Comment:
What are the proposed semantics of this field? Are para-IDs ordered? Is there a limit on the length of the vector? Can a `ParaId` appear multiple times, e.g. `[A, A, B]` to accommodate a (2/3, 1/3) split between A and B?

Reply:
If we want to accommodate more complex multiplicities, e.g. a 10-way split with 3/10 and 7/10, we would save space and time with a `Vec<(ParaId, u16)>`.

Reply:
This is intended to be the approximate split between blocks on which the given core is used. For example `[A, B, C]` would be a simple three-way split between paras A, B and C.

`[A, A, B]` would indeed correspond to 2/3 A and 1/3 B.

`[A, A]` is strictly equivalent to `[A]`.

Reply:
> If we want to accommodate more complex multiplicities, e.g. a 10-way split with 3/10 and 7/10, we would save space and time with a `Vec<(ParaId, u16)>`.

Yes, in the case that it's a 3:7 split, this would indeed occupy less space/time. Of course for the case of a single Para or equal sharing of the core it would increase the space/time.

The question is only what we want to optimise for.

Reply:
If we assume that the average case is biased sharing of a core, then we should definitely switch to a weighted vec. I really can't say for sure how I can imagine the typical usage but including a weight won't cost very much in terms of space and yet will save space in all circumstances where there is bias, so I guess we may as well do it.

Reply:
Done now.